### ISSUE TITLE
Installing and configuring OpenUpgrade

### ISSUE THREAD

**Issue by Yenthe666** [OCA/OpenUpgrade#434]
Hi guys,

I've been watching this project for a while and I'd now like to see how this works. I will have an internee in two weeks, which will be heavily creating and improving the migration processes for Odoo.
I'm trying to setup a base to fast forward him and to get the maximum out of his period but I run in to a few flaws / problems.

In the documentation (https://doc.therp.nl/openupgrade/analyse.html) I see the line "Set up two OpenUpgrade servers of subsequent Odoo releases", which really lacks information.
What I personally did is download and install an Odoo 8 and Odoo 9. I then downloaded the 8.0 and 9.0 version of the module 'openupgrade_records' and the module 'openupgrade' and added it to the Odoo.
I then downloaded the dependencies:
`sudo easy_install openerp-client-lib`
`sudo pip install openupgradelib`

<b>Odoo V8:</b>
I now to go Settings and click on 'Generate records', but no records are created after this is done?

<b>Odoo V9:</b>
When I install the module on the Odoo V9 it will work fine and install.
I then click on the menuitem 'Install all modules' so that I have all modules available. 
I then click on 'Generate Records' and wait a while. After this is done I see no records under the menuitem 'Records' though, shouldn't there be any?
I then click on 'Comparison configuration' and create a configuration to connect to the V8. The test connection succeeds fine. When I click on perform analysis it will run but no results are shown?

**Comment by Yenthe666** [comment]
@pedrobaeza and @StefanRijnhart could you guys have a look at this?
A screenshot with the result:
<img src="http://i.imgur.com/eacPbEb.png" alt="Details"/>

Another question: when I get a good result how should I continue? It will print out all the differences between the V8 and V9 models/fields and what then?

**Comment by pedrobaeza** [comment]
You don't need to do the comparison by yourself. It's already done and the result is in the migrations directory of each module. I think it's more useful to make some migration scripts that repeating this job. Or maybe you're trying to analyze your own modules?

**Comment by StefanRijnhart** [comment]
Hi Yenthe, thank you for your interest in the project. You need to install the openupgrade branches of the project, not install the records module on Odoo itself. But what is your reason for performing the analysis yourself? For the core modules, the analysis is included in the distribution.

**Comment by Yenthe666** [comment]
@pedrobaeza and @StefanRijnhart I guess I took a wrong approach then.
In essence: I want to migrate one module from V8 to V9 (an already existing module) so that I know how the whole migration process works and how it should be done. I'll then learn my trainee this so that he can migrate all (or atleast a lot) of modules from V8 to V9.
All contributions will flow back to OCA.

So I shouldn't install a clean Odoo 8 and Odoo 9 but I should install the one from here? Is there any installation script / procedure for this on how to do it?
What should I do when I have the Odoo environments installed? If you invest some time to get me going I'll make sure a lot of progress flows back to OCA. :smile:

**Comment by pedrobaeza** [comment]
When you say "already existing module", do you mean one module from regular Odoo distribution (for example, sale, purchase and so on)?

**Comment by Yenthe666** [comment]
@pedrobaeza yes I mean from the regular Odoo distribution. For example the module sale from V8 to V9.

**Comment by pedrobaeza** [comment]
OK, then you have to take the `openupgrade_analysis.txt` file inside migrations folder (https://github.com/OCA/OpenUpgrade/blob/9.0/addons/sale/migrations/9.0.1.0/openupgrade_analysis.txt), and look what it puts. These are the hints to provide the migration scripts. These only says what changes from one version to another (removed and new fields, type changes, etc), but you have to translate what it means, seeing the code from both. For example, you can see that one field is removed and another one is new, and in reality what it happens is that the field has been renamed. Then, you have to call the rename column method of openupgradelib to perform the migration. Check each line of the file and tell me back any doubt.

**Comment by Yenthe666** [comment]
@pedrobaeza so if I take the file example you gave and take the first field:
`sale         / account.analytic.line    / so_line (many2one)            : NEW relation: sale.order.line`
This for example means there is a new many2one made named 'so_line' which links to the model account.analytic.line from the model sale.order? 

Now, when I look at the second line:
`sale         / account.invoice          / section_id (many2one)         : DEL relation: crm.case.section`
This means that the field section_id (which is a many2one) has been removed in Odoo 9 and that the field section_id can just be removed from the model?

So, how should I handle XML views and their changes?
and one final question: I now see the output that it should become, but where or how should I script this? I'm missing the bigger picture about how I should modify the SQL queries so that they change/upgrade a field name / type. Could you give me a bit more direction there too?

**Comment by Yenthe666** [comment]
@pedrobaeza I'm going to start and try with the module 'project_timesheet' since this is a rather small module.
Judging by another pull request I should make two files: `post-migration.py` and `pre-migration.py`?
You can view those here: https://github.com/Yenthe666/OpenUpgrade/tree/9.0/addons/project_timesheet/migrations/9.0.1.0

So from there on, how should I continue?

**Comment by pedrobaeza** [comment]
`pre-migration.py` is called before updating the module itself, and you should place there the things that should be done before you can lose data. For example, there's a column that is renamed, then you should rename that column in this script, because if not, you will lose your data, as Odoo drops the column (because it doesn't exist in the model definition), and creates the new one specified in the new model definition. But if you rename it, then Odoo detects that the column already exists and do nothing.

`post-migration.py` is for tasks that have to be done after the new data model is applied, and thus is run after the module update. For example, there's now a new field that links the invoice with the refund, so you have to try to fill this field with the proper values (in this case, only heuristic would serve).

For most cases, you need to combine some code on both scripts. For example, one field that was a many2one now is a many2many. You rename the column to `None` (special value that indicates to OpenUpgrade that it has to apply a special name) on the pre-migration script, and then on then post-migration, insert the value in the new many2many field, reading from the SQL column that it has been renamed (`openupgrade.get_legacy_name(<old_field_name>)`). 

openupgradelib has methods to help you deal with this kind of transformations: https://github.com/OCA/openupgradelib/blob/master/openupgradelib/openupgrade.py#L67-L96.

**Comment by pedrobaeza** [comment]
```
sale / account.analytic.line / so_line (many2one) : NEW relation: sale.order.line
```

This is read as the model `account.analytic.line` has a new field that is called `so_line` that points to `sale.order.line`. As this field is not required (it would say it if so), you can do nothing, but maybe by heuristics you can try to fill it to get a more complete migration. In this case, I guess that now when you generate an analytic account from a sale (probably the sale of recurring contracts), this field is filled with the origin sale order line, so in theory there's no possibility of having this, because is a new Odoo feature.

**Comment by pedrobaeza** [comment]
```
sale / account.invoice / section_id (many2one) : DEL relation: crm.case.section
```

This means that this field has been removed. Now there's no possibility of selecting a sales team in the invoices. This seems logical, as invoicing is not exactly sales. And it seems that there's no replacement for this feature. So nothing to do in theory, but you can put a comment in a file inside migrations folder that we usually put called "user_notes.txt". In this file, we annotate some comments and thoughts for being reading and taken into account. For this case, I would put: "If you have any custom module that relies on the sales team in the invoices, you will need to change it accordingly or restore this field in your custom module."

**Comment by pedrobaeza** [comment]
For views there's usually nothing to do, as all data is treated as `noupdate="0"`, being removed and renewed automatically.

There are some data although tagged with `noupdate="1"` that you should consider to include for being updated if they have changed from one version to the following.

**Comment by BrayhanJC** [comment]
Hola @pedrobaeza ,, quisiera saber si estoy haciendo algo mal con la migración que estoy haciendo en estos momentos, al momento de resturar la base de datos que se me genera gracias a este fragmento de codigo y haciendo uso del script de migración. El cual es el siguiente:

wget https://raw.githubusercontent.com/OCA/OpenUpgrade/HEAD/scripts/migrate.py
python migrate.py --config=/etc/odoo-server.conf --database=name_database --run-migrations=8.0,9.0


Se me crea una base de datos llamada name_database_migrated, esta la intento restaurar en la versión de odoo 8, pero me sale un error el cual no he podido solucionar. Escribo a usted con el fin de que se me pueda ayudar y sugerirme como son los pasos para tener una migración exito. Los modulos que estan instalados en la versión 8 son aplicaciones de odoo. Muchas gracias por su atención.

**Comment by AaronHForgeFlow** [comment]
I wanted to analyze a non core module but it seems no records are generated when clicking on generate records wizard. The analysis is showing empty results. Not sure what it's missing:
![image](https://user-images.githubusercontent.com/19620251/34252050-5b6fb9dc-e642-11e7-97f0-61f8c7d607e6.png)
